File: codeql-custom-queries-java/android/avoid-android-sdk-location/avoid-android-sdk-location.java
class LocationManager {}
interface LocationListener {}
class LocationRequest {}
class Criteria {}
class GpsStatus {}
class FusedLocationProviderClient {}
class LocationCallback {}

// ============================================================
// 🚫 Noncompliant - utilisation des classes Android location legacy
// ============================================================
class NoncompliantLocationFields {

    LocationManager manager   = null; // $ Alert
    LocationListener listener = null; // $ Alert
    LocationRequest request   = null; // $ Alert
    Criteria criteria         = null; // $ Alert
    GpsStatus gpsStatus       = null; // $ Alert
}

// ============================================================
// 🚫 Noncompliant - variables locales legacy
// ============================================================
class NoncompliantLocationLocals {

    void getLocation() {
        LocationManager mgr  = null; // $ Alert
        Criteria crit        = null; // $ Alert
    }
}

// ============================================================
// 🚫 Noncompliant - tous les types legacy dans une seule méthode
// ============================================================
class NoncompliantLocationAllTypes {

    void allLegacyTypes() {
        LocationManager manager   = null; // $ Alert
        LocationListener listener = null; // $ Alert
        LocationRequest request   = null; // $ Alert
        Criteria criteria         = null; // $ Alert
        GpsStatus gpsStatus       = null; // $ Alert
    }
}

// ============================================================
// ✅ Compliant - utilisation du Fused Location Provider
// ============================================================
class CompliantLocation {

    FusedLocationProviderClient fusedClient = null; // OK
    LocationCallback callback               = null; // OK
}

File: codeql-custom-queries-java/android/avoid-android-sdk-location.kt
// Stubs
class LocationManager
interface LocationListener
class LocationRequest
class Criteria
class GpsStatus
class FusedLocationProviderClient
class LocationCallback

// 🚫 Noncompliant - using Android SDK location classes
class NoncompliantLocation {

    val manager: LocationManager? = null    // $ Alert
    val listener: LocationListener? = null  // $ Alert
    val request: LocationRequest? = null    // $ Alert
    val criteria: Criteria? = null          // $ Alert
    val gpsStatus: GpsStatus? = null        // $ Alert

    fun getLocation() {
        val mgr: LocationManager? = null    // $ Alert
        val crit: Criteria? = null          // $ Alert
    }
}

// ✅ Compliant - using fused location provider
class CompliantLocation {

    val fusedClient: FusedLocationProviderClient? = null  // OK
    val callback: LocationCallback? = null                // OK
}